### PR TITLE
arch: arc: optimization for irq lock/unlock

### PR DESCRIPTION
**Pull request by vonhust**
When SMP is enabled, the irq_lock/unlock will get and
release a global spin lock, but the codes changed in this
commit only need to lock the local cpu.

No affect on uniprocessor, but optimizations for SMP case.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

### CODE REVIEW

File: drivers/timer/arcv2_timer0.c [hunk 1]
```diff
@@ -354,14 +354,14 @@ void sys_clock_disable(void)
 	unsigned int key;  /* interrupt lock level */
 	u32_t control; /* timer control register value */
 
-	key = irq_lock();
+	key = arch_irq_lock();
```
Comment (abrodkin) [inline drivers/timer/arcv2_timer0.c:357]: Even though replacement of `irq_lock()` with `arch_irq_lock()` is harmless here as this operation makes only sense for a local CPU but my question is why do we need to globally disable IRQs on the core here? Typically we do that to exclude possibility to re-enter a critical section. But OK what happens if we re-enter this code? Or rather how is it possible to get here multiple times if not from IRQ handler, but then do we really expect IRQ handler to disable the clock???

**Review by nashif — APPROVED** [review]

File: arch/arc/core/timestamp.c [hunk 2]
```diff
@@ -29,10 +29,10 @@ u64_t z_tsc_read(void)
 	u64_t t;
 	u32_t count;
 
-	key = irq_lock();
+	key = arch_irq_lock();
```
Comment (abrodkin) [inline arch/arc/core/timestamp.c:32]: Here I also don't see any reason to have global local (i.e. all IRQs on this core) IRQ disable at all. What are we trying to mitigate here? We don't deal with global variables, so it seems to be pretty re-entrant friendly.

As for theoretical possibility to get an IRQ in between these 2 ops and get incorrect reading... well we may as well get IRQ later and get yet not very precise reading returned.

File: arch/arc/core/irq_manage.c [hunk 3]
```diff
@@ -147,7 +147,7 @@ void z_irq_priority_set(unsigned int irq, unsigned int prio, u32_t flags)
 {
 	ARG_UNUSED(flags);
 
-	unsigned int key = irq_lock();
+	unsigned int key = arch_irq_lock();
```
Comment (abrodkin) [inline arch/arc/core/irq_manage.c:150]: Not really clear why do we need to disable all local IRQs here as well.

File: arch/arc/core/irq_manage.c [hunk 4]
```diff
@@ -95,10 +95,10 @@ void z_arc_firq_stack_set(void)
 
 void arch_irq_enable(unsigned int irq)
 {
-	unsigned int key = irq_lock();
+	unsigned int key = arch_irq_lock();
```
Comment (abrodkin) [inline arch/arc/core/irq_manage.c:98]: I'd move `arch_irq_lock()` from here right to the `z_arc_v2_irq_unit_irq_enable_set()` so it's clear why do we need it.

See what we have in `z_arc_v2_irq_unit_irq_enable_set()`
```c
void z_arc_v2_irq_unit_irq_enable_set(
	int irq,
	unsigned char enable
	)
{
	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, enable);
}
```

Obviously we don't want to re-enter it in between setting `_ARC_V2_IRQ_SELECT` and writing to `_ARC_V2_IRQ_SELECT`. Because in that case we'll discard the first IRQ we meant to select and simply enable "the second" IRQ twice.

**Review by abrodkin — CHANGES_REQUESTED** [review]
What's done here seems to be harmless but I'd prefer to even get some benefits due to re-consideration of touched places at all. I'm pretty sure we might do better, read-on :)

File: arch/arc/core/irq_manage.c (hunk 4 above)
Comment (ruuddw) [reply arch/arc/core/irq_manage.c:98]: There's a warning and rationale given in arcv2_irq_unit.h that disabling interrupts is responsibility of the caller. Rationale is performance in case of multiple calls to IRQ handling - don't know if that's happening in current code, or a more defensive programming like Alexey proposes and always lock in the low-level irq control fuctions makes more sense...

**Review by andyross — APPROVED** [review]

**Comment by vonhust** [comment]
@abrodkin @ruuddw , the changed codes are legacy , i don't know the reasons. I understand @abrodkin 's opinion, it requires more investigation and  then do a clean up.  In fact, it can be viewed as some kind of self-protection.

File: arch/arc/core/irq_manage.c (hunk 4 above)
Comment (vonhust) [reply arch/arc/core/irq_manage.c:98]: make sense, i will move it to z_arc_v2_irq_unit_irq_enable_set

**Comment by vonhust** [comment]
@abrodkin @ruuddw  I added one more commit based on the feedback.

**Comment by abrodkin** [comment]
@vonhust why don't you squash both commits in one? IMHO makes no sense to first replace functions we're going to get rid of and then nuke them in the following commit.

File: drivers/timer/arcv2_timer0.c [hunk 5]
```diff
@@ -351,18 +351,11 @@ u32_t z_timer_cycle_get_32(void)
...
-	/* disable interrupt generation */
-
 	control = timer0_control_register_get();
 	timer0_control_register_set(control & ~_ARC_V2_TMR_CTRL_IE);
 
-	irq_unlock(key);
-
 	/* disable interrupt in the interrupt controller */
 
 	irq_disable(IRQ_TIMER0);
```
Comment (ruuddw) [inline drivers/timer/arcv2_timer0.c:361]: I think feedback of Andy was: why not get rid of this function completely?
If left in, then please state in a comment the assumption that interrupts are disabled when calling.
If left in, maybe move up the irq_disable(IRQ_TIMER0) to avoid the timer interrupt itself causes problems?

**Comment by ruuddw** [comment]
I'm ok with moving the critical section protections down to the low-level timer code. But please update the comments that state assumption is higher layers do the locking then as well.

**Comment by vonhust** [comment]
@ruuddw  @abrodkin  I update this  PR again according to the feedbacks, pls review.

File: arch/arc/core/timestamp.c (hunk 2 above)
Comment (ruuddw) [reply arch/arc/core/timestamp.c:32]: Looks like a valid race condition to protect against to me - don't want tsc_read() go back in time due to a tick interrupt while calling.

File: drivers/interrupt_controller/intc_arcv2_irq_unit.c [hunk 6]
```diff
@@ -170,16 +167,16 @@ static int arc_v2_irq_unit_resume(struct device *dev)
 #endif
 	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_VECT_BASE, ctx.irq_vect_base);
 
-	status32 = z_arc_v2_aux_reg_read(_ARC_V2_STATUS32);
```
Comment (ruuddw) [inline drivers/interrupt_controller/intc_arcv2_irq_unit.c:173]: I think removing the status register 'restore' code here is ok (restored elseware, I assume?), but I suggest to put this in a separate PR as it's not related to irq lock/unlock.

File: drivers/timer/arcv2_timer0.c [hunk 7]
```diff
@@ -351,21 +351,9 @@ u32_t z_timer_cycle_get_32(void)
...
-
-	control = timer0_control_register_get();
-	timer0_control_register_set(control & ~_ARC_V2_TMR_CTRL_IE);
-
-	irq_unlock(key);
-
-	/* disable interrupt in the interrupt controller */
-
-	irq_disable(IRQ_TIMER0);
+	/* This function is only called in sys_reboot,
```
Comment (ruuddw) [inline drivers/timer/arcv2_timer0.c:354]: There's a __weak definition in sys_clock_init.c already, why not simply remove completely?

File: include/arch/arc/v2/arcv2_irq_unit.h [hunk 8]
```diff
@@ -122,23 +132,35 @@ void z_arc_v2_irq_unit_prio_set(int irq, unsigned char prio)
...
+	arch_irq_unlock(key);
 }
 
 #if defined(CONFIG_ARC_SECURE_FIRMWARE)
+/*
+ * @brief Configure the secure state of interrupt
+ *
+ * Configure the secure state of the specified interrupt
+ *
+ * @return 1 in interrupt/exception state; 0 not in
```
Comment (ruuddw) [inline include/arch/arc/v2/arcv2_irq_unit.h:144]: this looks like a copy-paste mistake, there is no return value.

File: include/arch/arc/v2/arcv2_irq_unit.h [hunk 9]
```diff
@@ -189,19 +215,38 @@ bool z_arc_v2_irq_unit_is_in_isr(void)
  *
  * @return N/A
  */
+static ALWAYS_INLINE
+void z_arc_v2_irq_unit_trigger_set(int irq, unsigned int trigger)
+{
+	unsigned int key = arch_irq_lock();
 
-void z_arc_v2_irq_unit_trigger_set(int irq, unsigned int trigger);
```
Comment (ruuddw) [inline include/arch/arc/v2/arcv2_irq_unit.h:193]: Please check intc_arcv2_irq_unit.c as well - that has an implementation of this and other functions as well that do not lock interrupts. (looks like a mistake introduced when the timer API was updated: duplicate inline implementation in the header).

File: include/arch/arc/v2/arcv2_irq_unit.h [hunk 10]
```diff
@@ -214,8 +259,16 @@ unsigned int z_arc_v2_irq_unit_trigger_get(int irq);
  *
  * @return N/A
  */
+static ALWAYS_INLINE
```
Comment (ruuddw) [inline include/arch/arc/v2/arcv2_irq_unit.h:262]: The comment above still states interrupts must be locked for this operation.

**Review by ruuddw — CHANGES_REQUESTED** [review]
In general ok, but few clean-ups to be done.

**Review by abrodkin — CHANGES_REQUESTED** [review]
I'm really sorry for getting in the way again but I really don't like current structure of changes.
In https://github.com/zephyrproject-rtos/zephyr/pull/23983/commits/bab48da4524bed6d413c31a79045ef6701201520 we do way too many things at once:
1. Swap `irq_lock()` with `arch_irq_lock()` in just 1 place.
2. Add `arch_irq_lock()` in tons of places
3. Make `z_arc_v2_irq_unit_int_eoi()` and `z_arc_v2_irq_unit_trigger_{set|get}()` always inlined
That makes this patch very hard to understand and barely bisectable... i.e. it introduces such amout of changes that even if we determine it breaks something it won't help much in understanding what leads to a breakage. Well, and should we want to back-port it to a stable branch chances are high it won't apply cleanly due to amount of the diff. I would suggest to split it in 3 separate patches in line with what I mentioned above.

In https://github.com/zephyrproject-rtos/zephyr/pull/23983/commits/bab48da4524bed6d413c31a79045ef6701201520 we also have 2 independent changes:
1.  Again swap of `irq_lock()` with `arch_irq_lock()` (why not merge it with the same change in the first patch?).
2. "Deprecation" of `sys_clock_disable()` which IMHO makes sense to do in a separate commit as it has nothing to do with logic of the previous changes in this commit.

**Comment by vonhust** [comment]
@ruuddw , the places you mentioned are changed. I removed the sys_clock_disable and make it into a separate commit. 
@abrodkin,  i put irq unit related codes in one commit,  don't think it's worthy to make it too fine-grained, as the commit msg said it's more like a overhaul.

**Review by ruuddw — APPROVED** [review]

**Review by abrodkin — APPROVED** [review]
Thanks for all the changes!
Now it looks much better.

**Review by tyrantlizard77 — APPROVED** [review]
